File: src/extension.ts
import * as vs from 'vscode'
import { Range, Position } from 'vscode'
import * as tmp from 'tmp'
import * as fs from "fs-promise"

const spawn = require('cross-spawn')

export function activate(context: vs.ExtensionContext) {
    let disposable = vs.commands.registerCommand('extension.fntms.format', () => {
        const editor = vs.window.activeTextEditor; if (!editor) return
        const text = editor.document.getText()

        vs.window.withProgress({
            title: "Format file...",
            location: vs.ProgressLocation.Window
        }, () => {
            return new Promise(resolver => {
                tmp.file({ postfix: ".fs" }, async (_, path) => {
                    await fs.writeFile(path, text)

                    const cmd = spawn("mono", [__dirname + "/../../fantomas/Fantomas.exe", path])
                    cmd.stdout.on('data', (data: any) => console.log(`stdout: ${data}`))
                    cmd.stderr.on('data', (data: any) => console.log(`stderr: ${data}`))
                    await waitForEnd(cmd)

                    const formated = await fs.readFile(path, "UTF-8")
                    editor.edit(edit => {
                        edit.replace(new Range(
                            new Position(0, 0),
                            new Position(editor.document.lineCount, 0)),
                            formated)
                        resolver(undefined)
                    })
                })
            })
        })
    })
    context.subscriptions.push(disposable)
}

const waitForEnd = (cmd: any) => new Promise<void>(resolve => cmd.on('close', resolve))

File: src/extension.fs
module Extension

open System
open Fable.Import.vscode
open Fable.Import.JS
open Fable.PowerPack

module G = Fable.Import.Node.Globals
module I = Infrastructure

let private executeAsync text =
    promise {
        try 
            let! path = I.createTemp { postfix = ".fs" }
            do! I.write path text
            do! I.shell "mono" [ G.__dirname + "/../../fantomas/Fantomas.exe"; path ]
            let! formated = I.read path
            do! I.replaceText formated
        with
        | e -> e.Message |> window.showErrorMessage |> ignore
    } |> unbox<PromiseLike<_>>

let activate(context : ExtensionContext) =
    let handleCommand () = 
        window.activeTextEditor 
        |> Option.map (fun x -> x.document.getText())
        |> Option.map (executeAsync >> I.withProgress "Format file...")

    commands.registerCommand("extension.fntms.format", handleCommand |> unbox<Func<_,_>>)
    |> context.subscriptions.Add
